controller/konnect/reconciler_generic.go

```diff
@@ -169,6 +170,26 @@ func (r *KonnectEntityReconciler[T, TEnt]) Reconcile(
 	} else if res.Requeue {
 		return res, nil
 	}
+	// If a type has a KongConsumer ref, handle it.
+	res, err = handleKongConsumerRef(ctx, r.Client, ent)
+	if err != nil {
+		if !errors.As(err, &ReferencedKongConsumerIsBeingDeleted{}) {
```

Comment:
We should also handle the consumer's `NotFound` here to make the cascading deletion of the credential work as expected. Otherwise, after removing a `KongConsumer`, a credential gets stuck on `Reconciler error - {"controller": "CredentialBasicAuth", "controllerGroup": "configuration.konghq.com", "controllerKind": "CredentialBasicAuth", "CredentialBasicAuth": {"name":"basic-auth-1","namespace":"default"}, "namespace": "default", "name": "basic-auth-1", "reconcileID": "3934eb68-137c-4298-805a-370f3fdd0054", "error": "can't get the referenced KongConsumer default/consumer1: KongConsumer.configuration.konghq.com \"consumer1\" not found"}`.

Another aspect is that the `Consumer` can be removed from Konnect without removing its children in advance (the credentials will be deleted in cascade) - that's a different behavior than in the `Route` - `Service` relation where you cannot delete a service without dropping the routes depending on it first. That makes the deletion of entities owned by a removed/being removed consumer not necessary.

I experimented a bit and prepared a patch that seems to work correctly:

```diff
diff --git a/controller/konnect/errors.go b/controller/konnect/errors.go
index 0f1c86c6..bc9442e6 100644
--- a/controller/konnect/errors.go
+++ b/controller/konnect/errors.go
@@ -46,3 +46,14 @@ type ReferencedKongConsumerIsBeingDeleted struct {
 func (e ReferencedKongConsumerIsBeingDeleted) Error() string {
        return fmt.Sprintf("referenced Kong Consumer %s is being deleted", e.Reference)
 }
+
+// ReferencedKongConsumerDoesNotExist is an error type that is returned when
+type ReferencedKongConsumerDoesNotExist struct {
+       Reference types.NamespacedName
+       Err       error
+}
+
+// Error implements the error interface.
+func (e ReferencedKongConsumerDoesNotExist) Error() string {
+       return fmt.Sprintf("referenced Kong Consumer %s does not exist: %v", e.Reference, e.Err)
+}
diff --git a/controller/konnect/reconciler_generic.go b/controller/konnect/reconciler_generic.go
index faf71354..12b5dad3 100644
--- a/controller/konnect/reconciler_generic.go
+++ b/controller/konnect/reconciler_generic.go
@@ -173,20 +173,33 @@ func (r *KonnectEntityReconciler[T, TEnt]) Reconcile(
        // If a type has a KongConsumer ref, handle it.
        res, err = handleKongConsumerRef(ctx, r.Client, ent)
        if err != nil {
-               if !errors.As(err, &ReferencedKongConsumerIsBeingDeleted{}) {
-                       return ctrl.Result{}, err
-               }
-
-               // If the referenced KongConsumer is being deleted (has a non zero deletion timestamp)
-               // then we remove the entity if it has not been deleted yet (deletion timestamp is zero).
-               // We do this because Konnect blocks deletion of entities like Credentials
-               // if they contain entities like Routes.
-               if ent.GetDeletionTimestamp().IsZero() {
-                       if err := r.Client.Delete(ctx, ent); err != nil {
-                               return ctrl.Result{}, fmt.Errorf("failed to delete %s: %w", client.ObjectKeyFromObject(ent), err)
+               consumerGoneOrBeingDeleted := errors.As(err, &ReferencedKongConsumerDoesNotExist{}) || errors.As(err, &ReferencedKongConsumerIsBeingDeleted{})
+               if consumerGoneOrBeingDeleted {
+                       // If the referenced KongConsumer is not found or is being deleted and the object is being deleted, remove the
+                       // finalizer and let the deletion proceed without trying to delete the entity from Konnect as the KongConsumer
+                       // deletion will (or already has - in case of the consumer being gone) take care of it on the Konnect side.
+                       if !ent.GetDeletionTimestamp().IsZero() {
+                               if controllerutil.RemoveFinalizer(ent, KonnectCleanupFinalizer) {
+                                       if err := r.Client.Update(ctx, ent); err != nil {
+                                               if k8serrors.IsConflict(err) {
+                                                       return ctrl.Result{Requeue: true}, nil
+                                               }
+                                               return ctrl.Result{}, fmt.Errorf("failed to remove finalizer %s: %w", KonnectCleanupFinalizer, err)
+                                       }
+                                       log.Debug(logger, "konnect finalizer removed as the owning KongConsumer is being deleted or is already gone", ent)
+                               }
+                       } else {
+                               // If the entity for unknown reason is not being deleted yet (it should be if the owning KongConsumer is gone
+                               // or is being deleted), we should delete it.
+                               if err := r.Client.Delete(ctx, ent); err != nil {
+                                       return ctrl.Result{}, fmt.Errorf("failed to delete %s: %w", client.ObjectKeyFromObject(ent), err)
+                               }
+                               log.Debug(logger, "entity deleted as the owning KongConsumer is being deleted or is already gone", ent)
                        }
                        return ctrl.Result{}, nil
                }
+
+               return ctrl.Result{}, err
        } else if res.Requeue {
                return res, nil
        }
@@ -776,6 +789,9 @@ func handleKongConsumerRef[T constraints.SupportedKonnectEntityType, TEnt constr
                ); errStatus != nil || res.Requeue {
                        return res, errStatus
                }
+               if k8serrors.IsNotFound(err) {
+                       return ctrl.Result{}, ReferencedKongConsumerDoesNotExist{Reference: nn, Err: err}
+               }
 
                return ctrl.Result{}, fmt.Errorf("can't get the referenced KongConsumer %s: %w", nn, err)
        }
```